src/target/riscv/riscv.c

```diff
@@ -3076,15 +3076,39 @@ static int riscv_read_memory(struct target *target, target_addr_t address,
 		return ERROR_OK;
 	}
 
-	target_addr_t physical_addr;
-	int result = target->type->virt2phys(target, address, &physical_addr);
-	if (result != ERROR_OK) {
-		LOG_TARGET_ERROR(target, "Address translation failed.");
-		return result;
+	const target_addr_t page_offset_mask = RISCV_PGSIZE - 1;
+	const bool crosses_page_boundary =
+		((address + size * count - 1) & ~page_offset_mask) != (address & ~page_offset_mask);
+	if (address % size && crosses_page_boundary) {
+		int mmu_enabled;
+		int result = riscv_mmu(target, &mmu_enabled);
+		if (result != ERROR_OK)
+			return result;
+		if (mmu_enabled) {
+			LOG_TARGET_ERROR(target, "Accessing an element across a page boundary is not supported.");
+			return ERROR_FAIL;
+		}
```

Comment:
This code is used in both `riscv_read_memory()` and `riscv_write_memory()`.
Maybe it can be separated into a function like `int detect_write_on_page_boundary(...)`?

Reply:
Agree.
In fact, the entire function code is very similar. But I'm not sure if it's worth creating something like` riscv_rw_memory(..., bool is_write)` ， then both` riscv_read_memory()` and `riscv_write_memory()` can call it.

Reply:
>  But I'm not sure if it's worth creating something like `riscv_rw_memory(..., bool is_write)`

IMHO that is a great idea and will make the code much more maintainable.